man/dminit.1
.TH "DMINIT" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dminit \(en creates and initialises DMPACK databases
.SH SYNOPSIS
\f[B]dminit\f[R] \-\-help
.PP
\f[B]dminit\f[R] \-\-version
.PP
\f[B]dminit\f[R] \-\-\f[B]type\f[R] \f[I]type\f[R]
\-\-\f[B]database\f[R] \f[I]file\f[R] [\-\-\f[B]force\f[R]]
[\-\-\f[B]sync\f[R]] [\-\-\f[B]transfer\f[R]] [\-\-\f[B]wal\f[R]]
.SH DESCRIPTION
The \f[B]dminit\f[R] utility program creates beat, image, log, and
observation databases.
No action is performed if the specified database already exists, unless
command\-line argument \f[CR]\-\-force\f[R] is passed.
.PP
A synchronisation table is required for observation and log
synchronisation with an RPC server.
The argument may be omitted if this functionality is not used.
.PP
Write\-Ahead Logging (WAL) should be enabled for databases with multiple
readers.
.SH OPTIONS
.TP
\f[B]\-\-database\f[R], \f[B]\-d\f[R] \f[I]file\f[R]
File path of the new SQLite database.
.TP
\f[B]\-\-force\f[R], \f[B]\-F\f[R]
Force creation of tables even if the database already exists.
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-sync\f[R], \f[B]\-s\f[R]
Add synchronisation tables.
Enable for data synchronisation between client and server.
.TP
\f[B]\-\-transfer\f[R], \f[B]\-x\f[R]
Add transfers table to database if type is \f[CR]image\f[R].
.TP
\f[B]\-\-type\f[R], \f[B]\-t\f[R] [beat|image|log|observ]
Type of database, either beat, image, log, or observation.
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.TP
\f[B]\-\-wal\f[R], \f[B]\-W\f[R]
Enable Write\-Ahead Logging.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Database was created.
.TP
\f[B]1\f[R]
Failure.
Database creation failed.
.SH EXAMPLE
Create a beat database:
.IP
.EX
$ dminit \-t beat \-d beat.db \-W
.EE
.PP
Create a log database with remote synchronisation tables:
.IP
.EX
$ dminit \-t log \-d log.db \-s \-W
.EE
.PP
Create an observation database with remote synchronisation tables:
.IP
.EX
$ dminit \-t observ \-d observ.db \-s \-W
.EE
.PP
Create an image database with transfer table:
.IP
.EX
$ dminit \-t image \-d image.db \-x
.EE
.SH SEE ALSO
\f[I]dmbackup(1)\f[R], dmdbctl(1), \f[I]dmsync(1)\f[R]


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` error codes
* `Added` _fortran-nng_ interface bindings to NNG to `vendor/`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_group` for observation groups
* `Added` module `dm_ipc` for socket access
* `Added` module `dm_ipc_async` for asynchronous I/O
* `Added` module `dm_ipc_disco` for service discovery
* `Added` module `dm_ipc_message` for message handling
* `Added` module `dm_ipc_mutex` for mutex access
* `Added` module `dm_ipc_thread` for thread access
* `Added` module `dm_ipc_type` of IPC enumerators
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_mqueue_util`, `dm_pipe`,
  `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` database schema due to new data models
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` model of observation data structure `dm_observ`
* `Deleted` model of request data strucutre `dm_request`

## Programs

* `Changed` response value units in `dmved`

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)
* `Added` module `dm_linux` for system status access on Linux
* `Added` module `dm_modbus` for Modbus RTU/TCP connectivity (libmodbus)
* `Added` module `dm_ve` for Victron Energy VE.Direct protocol support
* `Added` library build date and system type parameters (CPP, FPP)
* `Added` Linux and FreeBSD abstractions in module `dm_system`
* `Changed` refactored database abstraction layer
* `Changed` refactored database schema
* `Changed` refactored command-line argument parsing
* `Changed` refactored logical function names
* `Fixed` wrong baud rate in GeoCOM class

### Programs

* `Added` program `dmbot` for remote control of clients through XMPP
* `Added` program `dmdwd` for Deutscher Wetterdienst (DWD) API monitoring
* `Added` program `dmmb` for sensor monitoring through Modbus RTU/TCP
* `Added` program `dmmbctl` for Modbus RTU/TCP command-line control
* `Added` program `dmsystem` for system monitoring
* `Added` program `dmved` for MPPT and battery monitoring (VE.Direct)
* `Added` interactive map view to `dmweb` (Leaflet)

### Documentation


app/dminit.f90
! dminit.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dminit
    !! Utility program that creates a beat, log, or observation database. No
    !! action is performed if the specified database already exists, unless
    !! command-line argument `--force` is passed. WAL mode should be enabled
    !! for databases with multiple readers. A synchronisation table is required
    !! for observation and log synchronisation with an DMPACK server, and can
    !! be omitted otherwise. The program also sets the database application id
    !! and user version.
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dminit'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    type :: app_type
        !! Command-line arguments.
        character(FILE_PATH_LEN) :: database = ' '       !! Path to database.
        integer                  :: type     = TYPE_NONE !! Type of database.
        logical                  :: force    = .false.   !! Force creation.
        logical                  :: sync     = .false.   !! Sync flag.
        logical                  :: transfer = .false.   !! Transfer flag.
        logical                  :: wal      = .false.   !! WAL flag.
    end type app_type

    integer        :: rc  ! Return code.
    type(app_type) :: app ! App settings.

    ! Initialise DMPACK.
    call dm_init()

    ! Get command-line arguments.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Create selected database type.
    rc = create_db(app%type, app%database, app%wal, app%sync, app%transfer)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)
contains
    integer function create_db(type, path, wal, sync, transfer) result(rc)
        !! Creates database schema.
        integer,      intent(in) :: type     !! Database type.
        character(*), intent(in) :: path     !! Path to database.
        logical,      intent(in) :: wal      !! WAL mode.
        logical,      intent(in) :: sync     !! Add sync tables.
        logical,      intent(in) :: transfer !! Add transfers table.

        type(db_type) :: db

        ! Open and create database.
        rc = dm_db_open(db, path, create=.true., wal=wal)

        if (dm_is_error(rc)) then
            call dm_error_out(rc, 'failed to create database')
            return
        end if

        ! Create tables.
        select case (type)
            case (TYPE_BEAT);   rc = dm_db_table_create_beats(db)
            case (TYPE_IMAGE);  rc = dm_db_table_create_images(db, sync=sync, transfer=transfer)
            case (TYPE_LOG);    rc = dm_db_table_create_logs(db, sync=sync)
            case (TYPE_OBSERV); rc = dm_db_table_create_observs(db, sync=sync)
        end select

        call dm_db_close(db)
        if (dm_is_error(rc)) call dm_error_out(rc, 'failed to create database')
    end function create_db

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments.
        type(app_type), intent(out) :: app !! App type.

        character(TYPE_NAME_LEN) :: type_name
        type(arg_parser_class)   :: parser

        call parser%add('type',     short='t', type=ARG_TYPE_STRING,   required=.true.) ! -t, --type [beat|log|observ]
        call parser%add('database', short='d', type=ARG_TYPE_DATABASE, required=.true.) ! -d, --database <path>
        call parser%add('force',    short='F', type=ARG_TYPE_LOGICAL)                   ! -F, --force
        call parser%add('sync',     short='s', type=ARG_TYPE_LOGICAL)                   ! -s, --sync
        call parser%add('transfer', short='x', type=ARG_TYPE_LOGICAL)                   ! -x, --transfer
        call parser%add('wal',      short='W', type=ARG_TYPE_LOGICAL)                   ! -W, --wal

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        ! Database type (observ, log, beat).
        call parser%get('type',     type_name)
        call parser%get('database', app%database)
        call parser%get('force',    app%force)
        call parser%get('sync',     app%sync)
        call parser%get('transfer', app%transfer)
        call parser%get('wal',      app%wal)

        app%type = dm_type_from_name(type_name)

        ! Validate options.
        rc = validate(app)
    end function read_args

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(inout) :: app !! App type.

        rc = E_INVALID

        select case (app%type)
            case (TYPE_OBSERV, TYPE_LOG, TYPE_BEAT)
                if (app%transfer) then
                    call dm_error_out(rc, 'argument --transfer is not allowed')
                    return
                end if

            case (TYPE_IMAGE)
                continue

            case default
                call dm_error_out(rc, 'argument --type is invalid (must be observ, log, beat, or image)')
                return
        end select

        rc = E_EXIST

        if (.not. app%force .and. dm_file_exists(app%database)) then
            call dm_error_out(rc, 'database ' // trim(app%database) // ' exists')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_db_version(.true.)
    end subroutine version_callback
end program dminit
